crates/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -0,0 +1,360 @@
+use std::collections::BTreeMap;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use chrono::{DateTime, Utc};
+use flox_types::version::Version;
+use serde::{Deserialize, Serialize};
+use serde_with::{DeserializeFromStr, SerializeDisplay};
+use thiserror::Error;
+
+use super::path_environment::{Original, PathEnvironment};
+use super::{copy_dir_recursive, PathPointer, ENV_DIR_NAME};
+use crate::providers::git::{GitCommandProvider, GitProvider};
+
+const GENERATIONS_METADATA_FILE: &str = "metadata.json";
+
+pub struct ReadOnly {}
+
+pub struct ReadWrite {
+    /// A base directory that contains the tempdir
+    /// of the checked out generations branch
+    /// as well as tempdirs created by [PathEnvironment] transactions.
+    tempdir_base: PathBuf,
+}
+
+/// A representation of the generations of an environment
+///
+/// Example File layout:
+///
+/// ./
+/// ├── 1
+/// │  └── env
+/// │     ├── manifest.toml
+/// │     └── manifest.lock
+/// ├── 2
+/// │  └── env
+/// │    └── manifest.toml (lockfile is optional)
+/// ├── ... N
+/// │  └── env
+/// │     └── manifest.toml
+/// └── metadata.json
+pub struct Generations<State> {
+    /// A floxmeta repository/branch that contains the generations of an environment
+    ///
+    /// - When [ReadOnly], this is assumend to be bare, but it is not enforced.
+    /// - When [ReadWrite], this is required to not be bare.
+    ///   This is enforced when created using [Self::writable].
+    repo: GitCommandProvider,
+
+    /// A path pointer for the environment that will
+    /// be associated with a realized generation
+    pointer: PathPointer,
+
+    branch: String,
+
+    state: State,
+}
+
+impl<S> Generations<S> {
+    /// Read the generations metadata for an environment
+    pub fn metadata(&self) -> Result<Metadata, GenerationsError> {
+        read_metadata(&self.repo, &self.branch)
+    }
+}
+
+impl Generations<ReadOnly> {
+    /// Create a new generations instance
+    pub fn new(
+        environment_repr: GitCommandProvider,
+        ref_name: String,
+        pointer: PathPointer,
+    ) -> Self {
+        Self {
+            repo: environment_repr,
+            branch: ref_name,
+            pointer,
+            state: ReadOnly {},
+        }
+    }
+
+    pub fn writable(
+        self,
+        tempdir_base: PathBuf,
+    ) -> Result<Generations<ReadWrite>, GenerationsError> {
+        let repo = checkout_to_tempdir(
+            &self.repo,
+            &self.branch,
+            tempfile::tempdir_in(&tempdir_base).unwrap().into_path(),
+        )?;
+
+        Ok(Generations {
+            repo,
+            branch: self.branch,
+            pointer: self.pointer,
+            state: ReadWrite { tempdir_base },
+        })
+    }
+}
+
+impl Generations<ReadWrite> {
+    /// Realize a generation as a [PathEnvironment]
+    ///
+    /// The generation can then be safely modified
+    /// and registered as a new generation using [Self::add_generation].
+    pub fn realize_generation(
```

Comment:
nit: I think this would be clearer
```
/// Return a [PathEnvironment] inside the `Generations` git repo
fn get_generation(generation: usize)

Reply:
updated the description and added some note about the tricky parts around that